crates/pink/sidevm/sidevm/src/net.rs

```diff
@@ -166,3 +167,34 @@ impl AsyncWrite for TcpStream {
         }
     }
 }
+
+impl Future for TcpConnector {
+    type Output = Result<TcpStream>;
+
+    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
+        use env::OcallError;
+        match ocall::poll_res(self.res_id.0) {
+            Ok(res_id) => Poll::Ready(Ok(TcpStream::new(ResourceId(res_id)))),
+            Err(OcallError::Pending) => Poll::Pending,
+            Err(err) => Poll::Ready(Err(err)),
+        }
+    }
+}
+
+impl TcpStream {
+    fn new(res_id: ResourceId) -> Self {
+        Self {
+            res_id,
+            buf: Default::default(),
+            start: 0,
+            filled: 0,
+        }
+    }
+
+    /// Initiate a TCP connection to a remote host.
+    pub async fn connect(addr: &str) -> Result<Self> {
+        let todo = "prevent local network probing";
```

Comment:
It's not a URL but a `hostname:port` pattern. We can resolve the hostname to IP address first in sidevm runtime and then make the TCP connection.
It's reasonable to do it on the host side to prevent the contract to scan services other than pruntime itself.
Do you think we should at least prevent contracts to access `localhost` here to protect pruntime itself?

Reply:
I think it's best to handle the connection in the host side. In the wasm, they can bypass the check by pointing a domain to 127.0.0.1 anyway